pkg/util/github/download_test.go

```diff
@@ -0,0 +1,224 @@
+			Expect(err).NotTo(Succeed())
+		})
+
+	})
+	var WorkPath = "./"
+	Context(("do DownloadAsset step 4"), func() {
+		BeforeEach(func() {
+			DeferCleanup(os.RemoveAll, "./"+github.DefaultWorkPath)
+			mux.HandleFunc("/repos/ow/re/releases", func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprint(w, `[
+					{
+						"id":1, 
+						"tag_name":"t", 
+						"name":"n", 
+						"assets": [{"id":1, "name":"a", "browser_download_url":"u"}]
+					}]`)
+			})
+		})
+
+		It("do get download url browser_download_url with unsupported proto scheme", func() {
+			ghClient, err := github.NewClientWithOption(&github.Option{
+				Owner: "ow",
+				Repo:  "re",
+			}, serverURL)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ghClient).NotTo(Equal(nil))
+			err = ghClient.DownloadAsset("t", "a", "f")
+			Expect(err).NotTo(Succeed())
+		})
+
+	})
+
+	Context(("do DownloadAsset step 4"), func() {
+		var filename = "f"
+		var downloadUrl = "/download"
+		BeforeEach(func() {
+			mux.HandleFunc("/repos/own/rep/releases", func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprintf(w, `[
+					{
+						"id":1, 
+						"tag_name":"t", 
+						"name":"n", 
+						"assets": [{"id":1, "name":"a", "browser_download_url":"%s"}]
+					}]`, serverURL+github.BaseURLPath+downloadUrl)
+			})
+			mux.HandleFunc(downloadUrl, func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprintf(w, "%s", "download content")
+			})
+		})
+
+		It("do get download url browser_download_url with supported proto scheme", func() {
+			ghClient, err := github.NewClientWithOption(&github.Option{
+				Owner:    "own",
+				Repo:     "rep",
+				WorkPath: WorkPath,
+			}, serverURL)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ghClient).NotTo(Equal(nil))
+			err = ghClient.DownloadAsset("t", "a", filename)
+			fmt.Println(err)
+			Expect(err).To(Succeed())
+		})
+
+		AfterEach(func() {
+			io.DeleteFile(filepath.Join(WorkPath, filename))
+		})
+	})
+})
+
+var _ = Describe("DownloadLatestCodeAsZipFile", func() {
+	var owner, repo, org, workPath = "owner", "repo", "org", "./"
+	Context(("do DownloadLatestCodeAsZipFile"), func() {
+
+		It("corrent url ", func() {
+			ghClient, err := github.NewClientWithOption(&github.Option{
+				Owner:    owner,
+				Repo:     repo,
+				Org:      org,
+				WorkPath: workPath,
+			}, serverURL)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ghClient).NotTo(Equal(nil))
+			err = ghClient.DownloadLatestCodeAsZipFile()
+			Expect(err).To(Succeed())
+		})
+
+		It("produce an error", func() {
+			ghClient, err := github.NewClientWithOption(&github.Option{
+				Owner:    owner,
+				Repo:     repo,
+				Org:      org,
+				WorkPath: "//",
+			}, serverURL)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ghClient).NotTo(Equal(nil))
+			err = ghClient.DownloadLatestCodeAsZipFile()
+			Expect(err).NotTo(Succeed())
+		})
+
+		AfterEach(func() {
+			DeferCleanup(io.DeleteFile, workPath+github.DefaultLatestCodeZipfileName)
+		})
+
+	})
```

Comment:
Thank you so much for your continued quality contributions lately! Especially the most, you've added some unit tests in a very clever way for almost the hardest parts of devstream to test. That's brilliant! 👍 @jxs1211 

Since I'm not particularly familiar with the github code, I haven't reviewed the substance of the code.

I noticed that you started your tests with golang's native tests, and wrote tests based on table structures. This would be the first time you've used ginkgo to write behavior-driven tests, so I briefly refactored this little piece of your code. I didn't mean to change your code, I just did my best to hopefully give you some inspiration. :heart: Just like what I learned about testing from @steinliber . Let's write code together and get better together!

```go
var _ = Describe("DownloadLatestCodeAsZipFile", func() {
	const (
		owner, repo, org             = "owner", "repo", "org"
		rightWorkPath, wrongWorkPath = "./", "//"
	)

	var (
		workPath string
		opts     *github.Option
	)

	JustBeforeEach(func() {
		opts = &github.Option{
			Owner:    owner,
			Repo:     repo,
			Org:      org,
			WorkPath: workPath,
		}
	})

	When("url is correct", func() {
		BeforeEach(func() {
			workPath = rightWorkPath
		})

		It("should return no error", func() {
			ghClient, err := github.NewClientWithOption(opts, serverURL)
			Expect(err).NotTo(HaveOccurred())
			Expect(ghClient).NotTo(Equal(nil))
			err = ghClient.DownloadLatestCodeAsZipFile()
			Expect(err).To(Succeed())
		})
	})

	When("the url is incorrect(caused by wrong work path)", func() {
		BeforeEach(func() {
			workPath = wrongWorkPath
		})

		It("should return an error", func() {
			ghClient, err := github.NewClientWithOption(opts, serverURL)
			Expect(err).NotTo(HaveOccurred())
			Expect(ghClient).NotTo(Equal(nil))
			err = ghClient.DownloadLatestCodeAsZipFile()
			Expect(err).NotTo(Succeed())
		})
	})

	AfterEach(func() {
		DeferCleanup(io.DeleteFile, workPath+github.DefaultLatestCodeZipfileName)
	})
})
```

Reply:
@aFlyBird0 I rewrote the test to try to make it more `behavior-driven`, I think it is more readable now, and the coverage increased to 100%.